cmd/sourced/cmd/web.go

```diff
@@ -32,54 +33,177 @@ func init() {
+// 2vudhlzztdlbg_gitbase_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /bin/dumb-init -- /bin/sh  ...   Up
+// 2vudhlzztdlbg_gitcollector_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /home/entrypoint.sh start- ...   Up                      0.0.0.0:10000->10000/tcp, 0.0.0.0:4040->4040/tcp, 8888/tcp
+// 2vudhlzztdlbg_gsc_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh postgres    Up                      0.0.0.0:5433->5432/tcp
+// 2vudhlzztdlbg_metadatadb_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh postgres    Up                      0.0.0.0:5432->5432/tcp
+// 2vudhlzztdlbg_postgres_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh redis ...   Up                      0.0.0.0:6379->6379/tcp
+// 2vudhlzztdlbg_redis_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /entrypoint.sh                   Up (health: starting)   0.0.0.0:8088->8088/tcp
+// 2vudhlzztdlbg_sourced-ui_1
+func runMonitor(ch chan<- error) {
+	runMonitorService := func(service string, ch chan<- error) {
+		for {
+			var stdout bytes.Buffer
+			if err := compose.RunWithIO(context.Background(),
+				os.Stdin, &stdout, nil, "ps", service); err != nil {
+				ch <- errors.Wrapf(err, "cannot get status service %s", service)
+				return
+			}
+
+			matches := stateExtractor.FindAllStringSubmatch(
+				normalizeNewLine(strings.TrimSpace(stdout.String())), -1)
+			for _, match := range matches {
+				state := match[1]
+
+				if strings.HasPrefix(state, "Exit") {
+					if service != "ghsync" && service != "gitcollector" {
+						ch <- fmt.Errorf("service '%s' is in state '%s'", service, state)
+						return
+					}
+
+					returnCode := state[len("Exit "):len(state)]
+					if returnCode != "0" {
+						ch <- fmt.Errorf("service '%s' exited with return code: %s",
+							service, returnCode)
+						return
+					}
+
+					continue
+				}
+
+				if state != "Up" {
+					ch <- fmt.Errorf("service '%s' is in state '%s'", service, state)
+					return
+				}
+
+			}
+
+			time.Sleep(1 * time.Second)
+		}
 	}
 
-	return nil
+	go func() {
+		var servicesBuf bytes.Buffer
+		if err := compose.RunWithIO(context.Background(),
+			os.Stdin, &servicesBuf, nil, "config", "--services"); err != nil {
+			ch <- errors.Wrap(err, "cannot get list of services")
+			return
+		}
+
+		services := strings.Split(normalizeNewLine(
+			strings.TrimSpace(servicesBuf.String())), "\n")
+
+		for _, service := range services {
+			go runMonitorService(service, ch)
+		}
+	}()
 }
 
 // OpenUI opens the browser with the UI.
 func OpenUI(timeout time.Duration) error {
+	var stdout bytes.Buffer
+	failFast, err := checkFailFast(&stdout)
```

Comment:
it took me a while to figure what is going on here. Do you think it would be easier to skip only error "container not available" instead? And fail for all other errors. Also, we might want to have the same check inside `waitForContainer`. Currently if, for example, docker server crashed we will ignore the error.

Reply:
> it took me a while to figure what is going on here. Do you think it would be easier to skip only error "container not available" instead? And fail for all other errors.

So you mean that we could avoid checking for those 3 types of errors? It makes sense to me. What do you think @dpordomingo? I'm asking you as you recently refactored this part regarding error handling AFAIR. I don't know whether there are other errors that we want to skip in addition to "container not available".

> Also, we might want to have the same check inside `waitForContainer`. Currently if, for example, docker server crashed we will ignore the error.

In this case, the monitor should correctly raise an error and exit.

Reply:
Currently, we are checking for 3 errors that we know how to display better. But the function can fail with many other errors we don't know about. They are just ignored now.

I propose to invert the logic. Check only for errors we want to skip and return everything else "fast" (including this 3 errors).

Maybe it's not necessary to do in this PR.

Reply:
> In this case, the monitor should correctly raise an error and exit.

that is true. But it's a side effect...

Reply:
> I propose to invert the logic. Check only for errors we want to skip and return everything else "fast" (including this 3 errors).

This makes sense. 👍 

> that is true. But it's a side effect...

Yeah, it's a side effect. The "problem" is that even if we add that check in the `waitForContainer` function, probably will be still the monitor to catch that error as it runs more frequently. But having that check even in `waitForContainer` will surely increase the understandability of the flow, so I'm ok adding it.